### PR TITLE
Feature / favorites, Test  / Category & Favorites, Refactor

### PR DESCRIPTION
**Pull request by JonasJallet**
> _Task description :_  
This pull request add favorite for user/place, refactor of user mock, test for category/place/user

- [x] should return all categories
- [x] should return one category by id
- [x] should create category and returns it
- [x] should throw error if trying to create category with same name
- [x] should delete category by id
- [x] should return updated category
- [x] should throw error if category does not exist

> _List of tests for place :_  
- [x] should throw error if trying to create place with same name and coordinates

> _List of tests for user :_  
- [x] should add a favorite place for the user
- [x] should throw error if user ID is invalid
- [x] should throw error if place ID is invalid
- [x] should remove a favorite place for the user

- [x] addFavorite
- [x] deleteFavorite

## Refactor💡
- [x] User Mock (so now we create 20 favorites foreach user)

- [X] yes -> recreate your database volume before use `make insert-mocks` command (show after your data with adminer)

### CODE REVIEW

File: back-end/src/entities/place.ts [hunk 1]
```diff
@@ -56,9 +63,8 @@ class Place extends BaseEntity {
   @Field(() => [Category])
   categories!: Category[];
 
-  @CreateDateColumn()
-  @Field()
-  createdAt!: Date;
+  @ManyToMany(() => User, (user) => user.favoritesPlaces)
+  users!: User[];
```
Comment (arnaudrenaud) [inline back-end/src/entities/place.ts:67]: Je la renommerais `usersWithFavorite` pour enlever toute ambiguïté sur la nature de la relation.

File: back-end/src/entities/user.ts [hunk 2]
```diff
@@ -53,11 +59,12 @@ class User extends BaseEntity {
...
-  places!: Place[];
+  ownedPlaces!: Place[];
 
-  @CreateDateColumn()
-  @Field()
-  createdAt!: Date;
+  @JoinTable({ name: "UserFavoritesPlaces" })
+  @ManyToMany(() => Place, (place) => place.users, { eager: true })
+  @Field(() => [Place])
+  favoritesPlaces!: Place[];
```
Comment (arnaudrenaud) [inline back-end/src/entities/user.ts:67]: Attention à l'accord de l'adjectif : `favoritePlaces` sans "s" à "favorite".

File: back-end/src/resolvers/userResolver.ts [hunk 3]
```diff
@@ -50,4 +51,28 @@ export class UserResolver {
...
+  async removeFavoritePlace(
+    @Arg("placeId") placeId: string,
+    @Ctx() { user }: Context
+  ): Promise<User> {
+    return User.deleteFavoritePlace((user as User).id, placeId);
+  }
+
+  @Authorized()
+  @Query(() => User)
+  async userFavoritesPlaces(@Ctx() { user }: Context): Promise<Place[]> {
```
Comment (arnaudrenaud) [inline back-end/src/resolvers/userResolver.ts:75]: Même faute d'accord : `userFavoritePlaces`.

File: back-end/src/tests/entities/category.test.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    it("should return updated category", async () => {
+      const createdCategory = await createNewCategory(newCategoriesDataset[2]);
+      const categoryId = createdCategory.id;
+      const partialCategory = { name: "updated-name" };
+      const updatedCategory = await Category.updateCategory(
+        categoryId,
+        partialCategory
+      );
+      expect(updatedCategory).toBeDefined();
+      expect(updatedCategory.name).toEqual(partialCategory.name);
```
Comment (arnaudrenaud) [inline back-end/src/tests/entities/category.test.ts:68]: Tu peux aussi tester que le nom a bien été mis à jour en base de donnée (en récupérant la valeur à jour avec `Category.getCategoryById`).

**Review by arnaudrenaud — CHANGES_REQUESTED** [review]

File: back-end/src/entities/user.ts (hunk 2 above)
Comment (JonasJallet) [reply back-end/src/entities/user.ts:67]: done

File: back-end/src/entities/place.ts (hunk 1 above)
Comment (JonasJallet) [reply back-end/src/entities/place.ts:67]: done

File: back-end/src/tests/entities/category.test.ts (hunk 4 above)
Comment (JonasJallet) [reply back-end/src/tests/entities/category.test.ts:68]: done

**Review by VinchC — APPROVED** [review]

**Review by JohannaStoicanescu — APPROVED** [review]
Good

**Review by PaulineHavez — APPROVED** [review]
Great job !!

**Pull request merged**